### PR TITLE
Optimize `println()` overloads added in P3142R0

### LINKED ISSUE #4630: Optimize P3142R0 Printing Blank Lines With `println()`
For followup after #4611 implementing #4524 is merged.

This PR is implementing the Standard's "Effects equivalent to" wording which is simple and unquestionably correct.

However, blank-line `println()` is likely to be a popular function to call, and it appears that it goes through a general codepath that does a lot of work:

https://github.com/microsoft/STL/blob/0515a05b394596de92d08cb0f352614479a2a883/stl/inc/print#L97-L117

The optimizer is unlikely to see all the way through this.

From a brief glance, it appears that we could provide the newline as a `string_view`, bypassing the construction of a `string _Unescaped_str`:

https://github.com/microsoft/STL/blob/0515a05b394596de92d08cb0f352614479a2a883/stl/inc/print#L111-L115

However, it is unclear to me whether we could further optimize this (without spending tons of complicated code). In particular, because we know that `"\n"` doesn't contain any exotic characters, could we simply call `_Print_noformat_nonunicode`?

For `<ostream>`, I believe that less thought is necessary:

https://github.com/microsoft/STL/blob/0515a05b394596de92d08cb0f352614479a2a883/stl/inc/ostream#L1246-L1263

`_STD _Print_noformat(_Ostr, _Unescaped_str)` eventually switches between Unicode and non-Unicode, but it has to construct `ostream::sentry` machinery first, so it wouldn't be worth digging any deeper.

### PR DESCRIPTION
**Pull request by frederick-vs-ja**
Fixes #4630.

- Add a function `__std_print_newline_only_to_unicode_console` to the import library to avoid potential branches around `WriteConsoleW`.
- Use `sputc` for non-Unicode new line printing for `ostream` to reduce potential branches.
- Use `fputc` for non-Unicode new line printing for `FILE*` as the function seems to be the simplest for the purpose in UCRT.

### CODE REVIEW

**Comment by StephanTLavavej** [comment]
I've pushed a conflict-free merge with main to pick up the 17.12p2 / Clang 18 toolset update. (This isn't usually critical, but it was a big reformatting PR.) No clang-format regeneration was necessary.

File: stl/inc/ostream [hunk 1]
```diff
@@ -187,6 +205,76 @@ ios_base::iostate _Print_noformat_unicode(ostream& _Ostr, const string_view _Str
     return _State;
 }
 
+template <class _Filebuf_type = filebuf>
+ios_base::iostate _Print_newline_only_unicode(ostream& _Ostr) {
```
Comment (StephanTLavavej) [inline stl/inc/ostream:209]: This is slightly modified from `_Print_noformat_unicode` above, and contains massive complicated machinery. We should extract the commonality.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks! :heart_eyes_cat: I pushed a conflict-free merge with `main` followed by several refactorings, shrinking the diff from +181 -2 to +115 -10. Originally, this PR looked like it was adding significant complexity, but now I like where it ended up. I'd still like @CaseyCarter to look at this.

**Review by CaseyCarter — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
Thanks for figuring out the surprisingly complicated logic needed to make these functions fast! :rocket: :heart_eyes_cat: :green_heart:

**Pull request merged**